Fix/custom option cache

machisuji opened the pull request:
WP [#26403](https://community.openproject.com/projects/openproject/work_packages/26403)

Technically one could of course just change the custom field's cache key.
For this we would need to perform a SQL query to check on the custom options every time the cache key is used, however. Which is a number of times for each custom field per request.

So I opted for this approach to just touch the custom field upon changes in the custom options in the one spot where a user can change them.

machisuji committed:
touch cf if options changed to invalidate cache

so that users get to see new options in filters.

The custom field's updated_at is not changed
automatically when anything in the has_many
relation changes. Hence we do this manually.

machisuji committed:
fixed deprecation, each_with_index will go in rails 5.1

ulferts commented:
@machisuji the specs are pointing out a bug in the production code.

machisuji committed:
don't fail when there are no options

ulferts commented on app/controllers/custom_fields_controller.rb:
```diff
@@ -63,9 +63,16 @@ def create
+
+      if last_update_after.try { |date| date > (last_update_before || date) }
+        @custom_field.touch # touch to invalidate cache
+      end
```
What you have build here should have been done by https://github.com/opf/openproject/blob/dev/app/models/custom_option.rb#L37 already. So the question is: Why didn't the standard way work?

ulferts committed:
fix touch: true option on custo_field association

Adding inverse_of on the custom_field association seems to fix a bug in rails which otherwise prevents the custom_field to be touched because of the after_commit callbacks defined on custom_field by acts_as_list
